presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSplitSource.java

```diff
@@ -254,6 +260,40 @@ public void testEmptyBucket()
         assertEquals(getSplits(hiveSplitSource, OptionalInt.of(3), 10).size(), 0);
     }
 
+    @Test
+    public void testLoadAllSplitsBeforeScheduling()
+            throws Exception
+    {
+        ConnectorSession session = new TestingConnectorSession(
+                new HiveSessionProperties(
+                        new HiveClientConfig().setLoadAllSplitsBeforeSchedulingForGroupedExecution(true),
+                        new OrcFileWriterConfig(),
+                        new ParquetFileWriterConfig()).getSessionProperties());
+        HiveSplitSource hiveSplitSource = HiveSplitSource.bucketed(
+                session,
+                "database",
+                "table",
+                TupleDomain.all(),
+                10,
+                10,
+                new DataSize(1, MEGABYTE),
+                new TestingHiveSplitLoader(),
+                Executors.newFixedThreadPool(5),
+                new CounterStat());
+        for (int i = 0; i < 10; i++) {
+            hiveSplitSource.addToQueue(new TestSplit(i, OptionalInt.of(0)));
+            assertEquals(hiveSplitSource.getBufferedInternalSplitCount(), i + 1);
+        }
+        Future<ConnectorSplitBatch> batch = hiveSplitSource.getNextBatch(new HivePartitionHandle(0), 10);
+        MILLISECONDS.sleep(200);
```

Comment:
Sleeps in tests usually result in flakiness.

Reply:
@nezihyigitbasi : For this one, we want it never be done, so it's not flaky -- although maybe we might not test what we want if 0.2s is too short. 

For the tests on line 292, it might be flaky. One thought is use similar mechanism as `waitUntilEquals` in `TestSqlTaskExecution`. Wondering if you have any other suggestions?

Reply:
@nezihyigitbasi I could port similar logic as `TestSqlTaskExecution::waitUntilEquals` to this test if you think it is valuable. If we decide to do that, we also need to change `testReaderWaitsForSplits` because it uses similar sleeping strategy as well.

Reply:
In the first assertion on L292 we check that the split batch is not ready 200ms after we call `getNextBatch`, instead of that how about we assert false until another thread calls `noMoreSplits`, and then assert true? I guess that's what we want -- we don't want the split batch to be ready until `noMoreSplits` is called. 

Also, why do you need the 800ms sleep after calling `noMoreSplits`? Is that call async? If yes, then that's flaky, and we should have some signaling mechanism to wait until that call is done.

Reply:
Done